src/main/java/com/sproutsocial/nsq/Publisher.java

```diff
@@ -1,103 +1,76 @@
+        Objects.requireNonNull(nsqd);
+
+        if (failoverNsqd == null) {
+            return (c, p) -> new SingleNsqdBalanceStrategy(c, p, nsqd);
+        } else {
+            return ListBasedBalanceStrategy.getFailoverStrategyBuilder(Arrays.asList(nsqd, failoverNsqd));
+        }
+    }
+
+    public Publisher(Client client, BiFunction<Client, Publisher, BalanceStrategy> balanceStrategyFactory) {
         super(client);
-        this.nsqd = HostAndPort.fromString(nsqd).withDefaultPort(4150);
-        this.failoverNsqd = failoverNsqd != null ? HostAndPort.fromString(failoverNsqd).withDefaultPort(4150) : null;
         client.addPublisher(this);
+        this.balanceStrategy = balanceStrategyFactory.apply(client, this);
     }
 
+
     public Publisher(String nsqd, String failoverNsqd) {
         this(Client.getDefaultClient(), nsqd, failoverNsqd);
     }
 
     public Publisher(String nsqd) {
-        this(nsqd, null);
-    }
-
-    @GuardedBy("this")
-    private void checkConnection() throws IOException {
-        if (con == null) {
-            if (isStopping) {
-                throw new NSQException("publisher stopped");
-            }
-            connect(nsqd);
-        }
-        else if (isFailover && Util.clock() - failoverStart > failoverDurationSecs * 1000) {
-            isFailover = false;
-            connect(nsqd);
-            logger.info("using primary nsqd");
-        }
+        this(Client.getDefaultClient(), nsqd, null);
     }
 
-    @GuardedBy("this")
-    private void connect(HostAndPort host) throws IOException {
-        if (con != null) {
-            con.close();
-        }
-        con = new PubConnection(client, host, this);
-        try {
-            con.connect(config);
-        }
-        catch(IOException e) {
-            con.close();
-            con = null;
-            throw e;
-        }
-        logger.info("publisher connected:{}", host);
-    }
 
     public synchronized void connectionClosed(PubConnection closedCon) {
-        if (con == closedCon) {
-            con = null;
-            logger.debug("removed closed publisher connection:{}", closedCon.getHost());
-        }
+        balanceStrategy.connectionClosed(closedCon);
     }
 
     public synchronized void publish(String topic, byte[] data) {
         checkNotNull(topic);
         checkNotNull(data);
         checkArgument(data.length > 0);
+        ConnectionDetails connectionDetails = balanceStrategy.getConnectionDetails();
         try {
-            checkConnection();
-            con.publish(topic, data);
-        }
-        catch (Exception e) {
-            logger.error("publish error with:{}", isFailover ? failoverNsqd : nsqd, e);
-            publishFailover(topic, data);
+            connectionDetails.getCon().publish(topic, data);
+        } catch (Exception e) {
+            connectionDetails.markFailure();
+            logger.error("publish error with", e);
+            publish(topic, data);
```

Comment:
I almost tripped on this recursive call. Didn't see a boundary condition there. Had to read the docs on `BalanceStrategy#getConnectionDetails()`.
I don't know how I feel about the distributed nature of these constructs (recursion and terminal condition). (NIT).

Reply:
And I am wondering again. Why would we not delegate all the publishing activity to the balance strategy?
I guess it is more of a philosophical question ATM. The retry has to happen somewhere and since Publisher is nothing but a delegate to the underlying pub-connection/connection_details/balance_strategy, why not delegate everything there, including the retry.

Reply:
Thats a fair point.  I chose this design to separate the logic of publishing from the logic of connection management.  

That said, the recursion is a bit of a concern.  To make it (maybe) worse, this is tail recursion, so a buggy BalanceStrategy could enter a forever loop inside a locked section.  That said, we do have test coverage on all the "out of connection" cases, so I think we are fine.  But the classes are open to a client injecting their own BalanceStrategy.

IMO, I think we are OK to leave as is.  If others want to inject their own, custom BalanceStrategy, they should test it thoroughly.  @blakesmith  do you have any thoughts on this one?